index.js

```diff
@@ -637,10 +666,47 @@ SendStream.prototype.sendFile = function sendFile(path) {
     fs.stat(p, function (err, stat) {
       if (err) return next(err)
       if (stat.isDirectory()) return next()
-      self.emit('file', p, stat)
-      self.send(p, stat)
+      checkPrecompressionAndSendFile(p, stat)
     })
   }
+
+  function checkPrecompressionAndSendFile(p, stat) {
+    self.emit('file', p, stat)
+    if (!self._precompressionFormats) return self.send(p, stat)
+
+    var state = {
+      contents: [],
+      extensionsToCheck: self._precompressionFormats.length
+    }
+
+    self._precompressionFormats.forEach(function (format) {
+      debug('stat "%s%s"', p, format.extension);
+      fs.stat(p + format.extension, function onstat(err, contentStat) {
```

Comment:
Always sending Vary means that it is also sent for uncompressable files such as images. Which means that caching proxies have to keep about 5-10 (the number of unique accept-encoding headers used by common browsers) versions of each image in cache instead of only one.

If served files are changed while server is running then the browser/proxy caches will in any case have stale data and missing Vary header for a file that was cached before it had precompressed variants is correct behavior.

I'm not familiar with good caching techniques in nodejs, but in any case I think caching cannot be enabled by default. It will cause wrong data to be sent out (Content-Length/Vary) if the files are changed while server is running - harming both development and production environments.

Maybe I (we?) should benchmark how much performance is lost by the extra stats?